### PR TITLE
Fixed a bug with falsey values within device meta updates

### PR DESCRIPTION
**Pull request by mjriley**
## Technical Summary
[Associated ticket](https://dimagi-dev.atlassian.net/browse/SAAS-14190)
This addresses a bug where the user's phone information did not line up with what we were displaying within our reports.
This fixes an issue with our update process for this data, as 'falsey' data would have previously been skipped -- i.e. a user who had previous had unsaved_forms would be unable to reset the number of unsaved_forms to 0, since 0 is falsey.

Apologies for bundling a refactor with the bug fix, rather than creating separate commits.

### Automated test coverage

New test suite in users/tests/test_models. Removed redundant tests from users/tests/test_user_model

### QA Plan

No QA required.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -751,19 +751,15 @@ def merge(self, other):
...
                 old_val = getattr(self, key)
-                if not old_val:
-                    setattr(self, key, new_val)
-                    continue
 
                 prop_is_date = isinstance(prop, DateTimeProperty)
-                if prop_is_date and new_val > old_val:
-                    setattr(self, key, new_val)
-                elif not prop_is_date and old_val != new_val:
-                    setattr(self, key, new_val)
```
Comment (esoergel) [inline corehq/apps/users/models.py:766]: What about this logic?  Do you know why the old version skipped setting the val if it was the same?  Seems useless, but I wonder if there's some spy logic that only saves this if an attr has been set.  If so, this change would result in extra writes.  Seems like a bit of a stretch, just wanted to probe into the rationale behind the old version of the code.

File: corehq/apps/users/tests/test_models.py [hunk 2]
```diff
@@ -94,3 +95,117 @@ def setUp(self):
...
+        self.original_meta.merge(self.updates_meta)
+        self.assertEqual(self.original_meta.num_unsent_forms, 5)
+
+    def test_merges_new_dates(self):
+        self.updates_meta.last_heartbeat = self.current_time
+
+        self.original_meta.merge(self.updates_meta)
+        self.assertEqual(self.original_meta.last_heartbeat, self.current_time)
+
+    def test_does_not_overwrite_unspecified_properties(self):
```
Comment (esoergel) [inline corehq/apps/users/tests/test_models.py:134]: Did you mean to swap `updates_meta` and `original_meta` in this test?  Judging from the method name, I'd expect `original_meta` to be the one with an extra property.  That and this body is identical to `test_merges_new_properties`.

File: corehq/apps/users/models.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/users/models.py:766]: I don't think there's anything like that going on.

**Review by snopoke — APPROVED** [review]
Approval assuming @esoergel comment about the test isn't blocking

File: corehq/apps/users/models.py (hunk 1 above)
Comment (mjriley) [reply corehq/apps/users/models.py:766]: Thanks for keeping an eye out for these edge cases. `DeviceAppMeta` only inherits from `DocumentSchema`, which, as far as I'm aware, doesn't do anything special, and also...is a couch model, which we'll likely migrate away from. Without a comment specifically calling out that we're doing something for performance reasons, I think it's worth it to simplify the logic

File: corehq/apps/users/tests/test_models.py (hunk 2 above)
Comment (mjriley) [reply corehq/apps/users/tests/test_models.py:134]: Thanks, that was an oversight for sure

**Review by esoergel — APPROVED** [review]
